# Video Game Sales

`Matheus Raz (mrol@cin.ufpe.br)`

`João Paulo Lins (jplo@cin.ufpe.br)`

In [110]:
from IPython.display import display

import numpy as np
import pandas as pd
import warnings 
warnings.filterwarnings('ignore')
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [111]:
df = pd.read_csv('VideoGamesSalesWithRatings.csv')
df.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  
0          51.0          8       322.0  Nintendo      E  
1           NaN        NaN         NaN       NaN    NaN  
2          73.0        8.3       709.0  Nintendo      E  
3          73.0          8       192.0  Nintendo      E  
4           NaN        NaN         NaN       NaN    NaN

In [112]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16719 entries, 0 to 16718
Data columns (total 16 columns):
Name               16717 non-null object
Platform           16719 non-null object
Year_of_Release    16450 non-null float64
Genre              16717 non-null object
Publisher          16665 non-null object
NA_Sales           16719 non-null float64
EU_Sales           16719 non-null float64
JP_Sales           16719 non-null float64
Other_Sales        16719 non-null float64
Global_Sales       16719 non-null float64
Critic_Score       8137 non-null float64
Critic_Count       8137 non-null float64
User_Score         10015 non-null object
User_Count         7590 non-null float64
Developer          10096 non-null object
Rating             9950 non-null object
dtypes: float64(9), object(7)
memory usage: 2.0+ MB


In [113]:
df

Name Platform  \
0                                             Wii Sports      Wii   
1                                      Super Mario Bros.      NES   
2                                         Mario Kart Wii      Wii   
3                                      Wii Sports Resort      Wii   
4                               Pokemon Red/Pokemon Blue       GB   
5                                                 Tetris       GB   
6                                  New Super Mario Bros.       DS   
7                                               Wii Play      Wii   
8                              New Super Mario Bros. Wii      Wii   
9                                              Duck Hunt      NES   
10                                            Nintendogs       DS   
11                                         Mario Kart DS       DS   
12                           Pokemon Gold/Pokemon Silver       GB   
13                                               Wii Fit      Wii   
14                                    Kinect Adventures!     X360   
15                                          Wii Fit Plus      Wii   
16                                    Grand Theft Auto V      PS3   
17                         Grand Theft Auto: San Andreas      PS2   
18                                     Super Mario World     SNES   
19          Brain Age: Train Your Brain in Minutes a Day       DS   
20                         Pokemon Diamond/Pokemon Pearl       DS   
21                                      Super Mario Land       GB   
22                                   Super Mario Bros. 3      NES   
23                                    Grand Theft Auto V     X360   
24                           Grand Theft Auto: Vice City      PS2   
25                         Pokemon Ruby/Pokemon Sapphire      GBA   
26           Brain Age 2: More Training in Minutes a Day       DS   
27                           Pokemon Black/Pokemon White       DS   
28                                Gran Turismo 3: A-Spec      PS2   
29                        Call of Duty: Modern Warfare 3     X360   
...                                                  ...      ...   
16689              Help Wanted: 50 Wacky Jobs (jp sales)      Wii   
16690  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii   
16691                              Neo Angelique Special      PSP   
16692                      Outdoors Unleashed: Africa 3D      3DS   
16693                                          Real Rode      PS2   
16694                                     Pony Friends 2       PC   
16695                       Dynasty Warriors: Eiketsuden      PS3   
16696                  Metal Gear Solid V: Ground Zeroes       PC   
16697                                  PGA European Tour      N64   
16698                                     Carmageddon 64      N64   
16699                                    Planet Monsters      GBA   
16700                                             Breach       PC   
16701                                   Bust-A-Move 3000       GC   
16702                                   Mega Brain Boost       DS   
16703                              The Longest 5 Minutes      PSV   
16704                           Mezase!! Tsuri Master DS       DS   
16705  Eiyuu Densetsu: Sora no Kiseki Material Collec...      PSP   
16706                            STORM: Frontline Nation       PC   
16707                                   Strawberry Nauts      PSV   
16708                                           Plushees       DS   
16709                                            15 Days       PC   
16710                      Men in Black II: Alien Escape       GC   
16711                                   Aiyoku no Eustia      PSV   
16712                 Woody Woodpecker in Crazy Castle 5      GBA   
16713   SCORE International Baja 1000: The Official Game      PS2   
16714                      Samurai Warriors: Sanada Maru      PS3   
16715                                   LMA Manage

In [114]:
minor = df['Year_of_Release'].min()
major = df['Year_of_Release'].max()
print("Menor ano da base: %d"%(minor))
display(df[df['Year_of_Release'] == minor])
print("Maior ano da base: %d"%(major))
display(df[df['Year_of_Release'] == major])

Menor ano da base: 1980


Name Platform  Year_of_Release     Genre   Publisher  \
262         Asteroids     2600           1980.0   Shooter       Atari   
546   Missile Command     2600           1980.0   Shooter       Atari   
1764          Kaboom!     2600           1980.0      Misc  Activision   
1968         Defender     2600           1980.0      Misc       Atari   
2650           Boxing     2600           1980.0  Fighting  Activision   
4019       Ice Hockey     2600           1980.0    Sports  Activision   
5360          Freeway     2600           1980.0    Action  Activision   
6301           Bridge     2600           1980.0      Misc  Activision   
6876         Checkers     2600           1980.0      Misc       Atari   

      NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
262       4.00      0.26       0.0         0.05          4.31           NaN   
546       2.56      0.17       0.0         0.03          2.76           NaN   
1764      1.07      0.07       0.0         0.01          1.15           NaN   
1968      0.99      0.05       0.0         0.01          1.05           NaN   
2650      0.72      0.04       0.0         0.01          0.77           NaN   
4019      0.46      0.03       0.0         0.01          0.49           NaN   
5360      0.32      0.02       0.0         0.00          0.34           NaN   
6301      0.25      0.02       0.0         0.00          0.27           NaN   
6876      0.22      0.01       0.0         0.00          0.24           NaN   

      Critic_Count User_Score  User_Count Developer Rating  
262            NaN        NaN         NaN       NaN    NaN  
546            NaN        NaN         NaN       NaN    NaN  
1764           NaN        NaN         NaN       NaN    NaN  
1968           NaN        NaN         NaN       NaN    NaN  
2650           NaN        NaN         NaN       NaN    NaN  
4019           NaN        NaN         NaN       NaN    NaN  
5360           NaN        NaN         NaN       NaN    NaN  
6301           NaN        NaN         NaN       NaN    NaN  
6876           NaN        NaN         NaN       NaN    NaN

Maior ano da base: 2020


Name Platform  Year_of_Release       Genre Publisher  \
5936  Imagine: Makeup Artist       DS           2020.0  Simulation   Ubisoft   

      NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
5936      0.27       0.0       0.0         0.02          0.29           NaN   

      Critic_Count User_Score  User_Count Developer Rating  
5936           NaN        tbd         NaN   Ubisoft      E

In [115]:
df.drop(labels=5936,inplace=True)
major = df['Year_of_Release'].max()
print("Maior ano da base: %d"%(major))
display(df[df['Year_of_Release'] == major])

Maior ano da base: 2017


Name Platform  \
14086  Phantasy Star Online 2 Episode 4: Deluxe Package      PS4   
16222  Phantasy Star Online 2 Episode 4: Deluxe Package      PSV   
16385                  Brothers Conflict: Precious Baby      PSV   

       Year_of_Release         Genre     Publisher  NA_Sales  EU_Sales  \
14086           2017.0  Role-Playing          Sega       0.0       0.0   
16222           2017.0  Role-Playing          Sega       0.0       0.0   
16385           2017.0        Action  Idea Factory       0.0       0.0   

       JP_Sales  Other_Sales  Global_Sales  Critic_Score  Critic_Count  \
14086      0.04          0.0          0.04           NaN           NaN   
16222      0.01          0.0          0.01           NaN           NaN   
16385      0.01          0.0          0.01           NaN           NaN   

      User_Score  User_Count Developer Rating  
14086        NaN         NaN       NaN    NaN  
16222        NaN         NaN       NaN    NaN  
16385        NaN         NaN       NaN    NaN

O jogo `Imagine: Makeup Artist` possui vendas espalhadas pelo mundo todo mesmo antes de seu lançamento, isso nos levou a pensar que fosse um tipo de pre-venda, porém após algumas pesquisas constatamos que o jogo foi lançado em 2009 e não possuia nenhuma previsão de lançamento para 2020, por isso tudo indica que seja um dado ruidoso, nos levando a questionar se os outros valores referentes a ele estariam realmente corretos. Optamos então por deletar esse valor do nosso dataset.

In [116]:
df.drop(labels=[14086,16222,16385],inplace=True)
major = df['Year_of_Release'].max()
print("Maior ano da base: %d"%(major))
display(df[df['Year_of_Release'] == major])

Maior ano da base: 2016


Name Platform  \
94                                               FIFA 17      PS4   
108                                     Pokemon Sun/Moon      3DS   
171                           Uncharted 4: A Thief's End      PS4   
245                       Call of Duty: Infinite Warfare      PS4   
289                                        Battlefield 1      PS4   
325                            Tom Clancy's The Division      PS4   
571                                              FIFA 17     XOne   
640                       Call of Duty: Infinite Warfare     XOne   
708                                      Far Cry: Primal      PS4   
709                                        Battlefield 1     XOne   
758                            Tom Clancy's The Division     XOne   
767                                            Overwatch      PS4   
921                                             NBA 2K17      PS4   
968                                            Mafia III      PS4   
1007                                       Madden NFL 17      PS4   
1056                                        No Man's Sky      PS4   
1107                                      Dark Souls III      PS4   
1216                              Ratchet & Clank (2016)      PS4   
1275                                         Doom (2016)      PS4   
1439                                      Gears of War 4     XOne   
1466                                       Yokai Watch 3      3DS   
1475                         The Elder Scrolls V: Skyrim      PS4   
1615                                       Madden NFL 17     XOne   
1801            Naruto Shippuden: Ultimate Ninja Storm 4      PS4   
1806                                        Watch Dogs 2      PS4   
1904                                     Forza Horizon 3     XOne   
1948                                           Overwatch     XOne   
2012           The Legend of Zelda: Twilight Princess HD     WiiU   
2024                                            NBA 2K17     XOne   
2056                                   Pokken Tournament     WiiU   
...                                                  ...      ...   
16505                               River City Ransom SP      3DS   
16507                                  Shin Hayarigami 2      PS4   
16512                                             Ride 2     XOne   
16513                                             Ride 2       PC   
16530                                         Dungeons 2      PS4   
16534                            Carmageddon: Max Damage      PS4   
16539  Zettai Kaikyuu Gakuen: Eden with Roses and Pha...      PSV   
16548                        Assassin's Creed Chronicles     XOne   
16562  Naruto Shippuden: Ultimate Ninja Storm Collection      PS3   
16587                        Taishou x Alice: All in One      PSV   
16589                      Farming 2017 - The Simulation      PS4   
16594                                      Hitman (2016)     XOne   
16621              Codename: Panzers Complete Collection       PC   
16627                           Norn + Nonette: Act Tune      PSV   
16634                           Sébastien Loeb Rally Evo     XOne   
16647                                  Rugby Challenge 3     XOne   
16648                                 Eve: Burst Error R      PSV   
16649                                             ZombiU     XOne   
16656                                   Fast Racing Neo      WiiU   
16664                                   Prison Architect     XOne   
16676      Metal Gear Solid V: The Definitive Experience     XOne   
16681                   Aikatsu Stars! My Special Appeal      3DS   
16684                               Kyoukai no Shirayuki      PSV   
16688                                          7'scarlet      PSV   
16695                       Dynasty Warriors: Eiketsuden      PS3   
16703                              The Longest 5 Minutes      PSV   
16707                                   Strawberry

Os jogos `Phantasy Star Online 2 Episode 4: Deluxe Package` de PS4 e PSV e `Brothers Conflict: Precious Baby` foram lançados em 2016 apenas no Japão e tinham como 2017 o ano de lançamento no resto do mundo. Porém, como a base data de 22 de Dezembro de 2016, não foi computado as vendas no resto do mundo, nem nota da crítica e etc. Sendo assim, optamos por deletar esses valores do nosso dataset.

In [117]:
df.columns

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales', 'Critic_Score',
       'Critic_Count', 'User_Score', 'User_Count', 'Developer', 'Rating'],
      dtype='object')

In [118]:
qtdUserScoreTbd = len(list(filter(lambda x : x is True, [x == 'tbd'  for x in df['User_Score']])))
display("Quantidade de valores \"tbd\" na coluna User_Score: %d"%(qtdUserScoreTbd))

df['User_Score'].replace('tbd', np.nan, inplace=True)

qtdUserScoreNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['User_Score']])))
display("Quantidade de valores \"NaN\" na coluna User_Score após a converção: %d"%(qtdUserScoreNan))

df['User_Score'] = df['User_Score'].apply(lambda x : float(x))
df.info()

'Quantidade de valores "tbd" na coluna User_Score: 2424'

'Quantidade de valores "NaN" na coluna User_Score após a converção: 9125'

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16715 entries, 0 to 16718
Data columns (total 16 columns):
Name               16713 non-null object
Platform           16715 non-null object
Year_of_Release    16446 non-null float64
Genre              16713 non-null object
Publisher          16661 non-null object
NA_Sales           16715 non-null float64
EU_Sales           16715 non-null float64
JP_Sales           16715 non-null float64
Other_Sales        16715 non-null float64
Global_Sales       16715 non-null float64
Critic_Score       8137 non-null float64
Critic_Count       8137 non-null float64
User_Score         7590 non-null float64
User_Count         7590 non-null float64
Developer          10095 non-null object
Rating             9949 non-null object
dtypes: float64(10), object(6)
memory usage: 2.2+ MB


A coluna `User_Score` possuia valores igual a `tbd` que no Metacritic significa "to be determined". Ou seja, é quando não houve o número mínimo de 4 avaliações para que o Metacritic pode-se considerar a avaliação média dos usuários. Sendo assim, resolvemos considerar que os valores iguais a `tbd` equivalem a `NaN`, visto que os valores correspondentes a `User_count` nesses registros está como `NaN`, o que implíca dizer que não houve usuários avaliadores.

Após a substituição de valores, convertemos o tipo da coluna `User_Score` de categórico para numérico

In [119]:
qtdAnosNan = len(df[np.isnan(df['Year_of_Release'])])
display("Quantidade de valores \"NaN\" na coluna Year_of_Release: %d"%(qtdAnosNan))

qtdNASalesNan = len(df[np.isnan(df['NA_Sales'])])
display("Quantidade de valores \"NaN\" na coluna NA_Sales: %d"%(qtdNASalesNan))

qtdEUSalesNan = len(df[np.isnan(df['EU_Sales'])])
display("Quantidade de valores \"NaN\" na coluna EU_Sales: %d"%(qtdEUSalesNan))

qtdJPSalesNan = len(df[np.isnan(df['JP_Sales'])])
display("Quantidade de valores \"NaN\" na coluna JP_Sales: %d"%(qtdJPSalesNan))

qtdOtherSalesNan = len(df[np.isnan(df['Other_Sales'])])
display("Quantidade de valores \"NaN\" na coluna Other_Sales: %d"%(qtdOtherSalesNan))

qtdGlobalSalesNan = len(df[np.isnan(df['Global_Sales'])])
display("Quantidade de valores \"NaN\" na coluna Global_Sales: %d"%(qtdGlobalSalesNan))

qtdCriticScoreNan = len(df[np.isnan(df['Critic_Score'])])
display("Quantidade de valores \"NaN\" na coluna Critic_Score: %d"%(qtdCriticScoreNan))

qtdCriticCountNan = len(df[np.isnan(df['Critic_Count'])])
display("Quantidade de valores \"NaN\" na coluna Critic_Count: %d"%(qtdCriticCountNan))

qtdUserScoreNan = len(df[np.isnan(df['User_Score'])])
display("Quantidade de valores \"NaN\" na coluna User_Score: %d"%(qtdUserScoreNan))

qtdUserCountNan = len(df[np.isnan(df['User_Count'])])
display("Quantidade de valores \"NaN\" na coluna User_Count: %d"%(qtdUserCountNan))

'Quantidade de valores "NaN" na coluna Year_of_Release: 269'

'Quantidade de valores "NaN" na coluna NA_Sales: 0'

'Quantidade de valores "NaN" na coluna EU_Sales: 0'

'Quantidade de valores "NaN" na coluna JP_Sales: 0'

'Quantidade de valores "NaN" na coluna Other_Sales: 0'

'Quantidade de valores "NaN" na coluna Global_Sales: 0'

'Quantidade de valores "NaN" na coluna Critic_Score: 8578'

'Quantidade de valores "NaN" na coluna Critic_Count: 8578'

'Quantidade de valores "NaN" na coluna User_Score: 9125'

'Quantidade de valores "NaN" na coluna User_Count: 9125'

In [120]:
qtdNameNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Name']])))
display("Quantidade de valores \"NaN\" na coluna Name: %d"%(qtdNameNan))

qtdPlatformNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Platform']])))
display("Quantidade de valores \"NaN\" na coluna Platform: %d"%(qtdPlatformNan))

qtdGenreNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Genre']])))
display("Quantidade de valores \"NaN\" na coluna Genre: %d"%(qtdGenreNan))

qtdPublisherNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Publisher']])))
display("Quantidade de valores \"NaN\" na coluna Publisher: %d"%(qtdPublisherNan))

qtdDeveloperNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Developer']])))
display("Quantidade de valores \"NaN\" na coluna Developer: %d"%(qtdDeveloperNan))

qtdRatingNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Rating']])))
display("Quantidade de valores \"NaN\" na coluna Rating: %d"%(qtdRatingNan))

'Quantidade de valores "NaN" na coluna Name: 2'

'Quantidade de valores "NaN" na coluna Platform: 0'

'Quantidade de valores "NaN" na coluna Genre: 2'

'Quantidade de valores "NaN" na coluna Publisher: 54'

'Quantidade de valores "NaN" na coluna Developer: 6620'

'Quantidade de valores "NaN" na coluna Rating: 6766'

In [121]:
df = df.dropna(subset = ['Name', 'Genre', 'Publisher', 'Year_of_Release'])

qtdNameNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Name']])))
qtdGenreNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Genre']])))
qtdPublisherNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Publisher']])))
qtdAnosNan = len(df[np.isnan(df['Year_of_Release'])])

display("Quantidade após drop desses valores \"NaN\" na coluna Name: %d"%(qtdNameNan))
display("Quantidade após drop desses valores \"NaN\" na coluna Genre: %d"%(qtdGenreNan))
display("Quantidade após drop desses valores \"NaN\" na coluna Publisher: %d"%(qtdPublisherNan))
display("Quantidade após drop desses valores \"NaN\" na coluna Year_of_Release: %d"%(qtdAnosNan))

'Quantidade após drop desses valores "NaN" na coluna Name: 0'

'Quantidade após drop desses valores "NaN" na coluna Genre: 0'

'Quantidade após drop desses valores "NaN" na coluna Publisher: 0'

'Quantidade após drop desses valores "NaN" na coluna Year_of_Release: 0'

In [122]:
qtdCriticScoreNan = len(df[np.isnan(df['Critic_Score'])])
display("Quantidade de valores \"NaN\" na coluna Critic_Score: %d"%(qtdCriticScoreNan))

qtdCriticCountNan = len(df[np.isnan(df['Critic_Count'])])
display("Quantidade de valores \"NaN\" na coluna Critic_Count: %d"%(qtdCriticCountNan))

qtdUserScoreNan = len(df[np.isnan(df['User_Score'])])
display("Quantidade de valores \"NaN\" na coluna User_Score: %d"%(qtdUserScoreNan))

qtdUserCountNan = len(df[np.isnan(df['User_Count'])])
display("Quantidade de valores \"NaN\" na coluna User_Count: %d"%(qtdUserCountNan))

df['Critic_Score'].replace(np.nan, 0, inplace=True)
df['Critic_Count'].replace(np.nan, 0, inplace=True)
df['User_Score'].replace(np.nan, 0, inplace=True)
df['User_Count'].replace(np.nan, 0, inplace=True)

display("Após substituição dos valores \"NaN\" para \"0\" temos:")

qtdCriticScoreZero = len(list(filter(lambda x : x is True, [x == 0  for x in df['Critic_Score']])))
display("Quantidade de valores \"0\" na coluna Critic_Score: %d"%(qtdCriticScoreZero))

qtdCriticCountZero = len(list(filter(lambda x : x is True, [x == 0  for x in df['Critic_Count']])))
display("Quantidade de valores \"0\" na coluna Critic_Count: %d"%(qtdCriticCountZero))

qtdUserScoreZero = len(list(filter(lambda x : x is True, [x == 0  for x in df['User_Score']])))
display("Quantidade de valores \"0\" na coluna User_Score: %d"%(qtdUserScoreZero))

qtdUserCountZero = len(list(filter(lambda x : x is True, [x == 0  for x in df['User_Count']])))
display("Quantidade de valores \"0\" na coluna User_Count: %d"%(qtdUserCountZero))

'Quantidade de valores "NaN" na coluna Critic_Score: 8430'

'Quantidade de valores "NaN" na coluna Critic_Count: 8430'

'Quantidade de valores "NaN" na coluna User_Score: 8951'

'Quantidade de valores "NaN" na coluna User_Count: 8951'

'Após substituição dos valores "NaN" para "0" temos:'

'Quantidade de valores "0" na coluna Critic_Score: 8430'

'Quantidade de valores "0" na coluna Critic_Count: 8430'

'Quantidade de valores "0" na coluna User_Score: 8952'

'Quantidade de valores "0" na coluna User_Count: 8951'

In [123]:
qtdDeveloperNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Developer']])))
display("Quantidade de valores \"NaN\" na coluna Developer: %d"%(qtdDeveloperNan))
        
def teste(x):
    if((x['Developer'] is np.nan) and (x['Publisher'] is not np.nan)):
        x['Developer'] = x['Publisher']
    return x
        
df[['Developer', 'Publisher']] = df[['Developer', 'Publisher']].apply(teste, axis=1)


qtdDeveloperNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Developer']])))
display("Quantidade de valores \"NaN\" na coluna Developer após substituição: %d"%(qtdDeveloperNan))

qtdRatingNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Rating']])))
display("Quantidade de valores \"NaN\" na coluna Rating: %d"%(qtdRatingNan))

df['Rating'].replace(np.nan, 'E', inplace=True)

qtdRatingNan = len(list(filter(lambda x : x is True, [x is np.nan  for x in df['Rating']])))
display("Quantidade de valores \"NaN\" na coluna Rating após substituição: %d"%(qtdRatingNan))

'Quantidade de valores "NaN" na coluna Developer: 6509'

'Quantidade de valores "NaN" na coluna Developer após substituição: 0'

'Quantidade de valores "NaN" na coluna Rating: 6646'

'Quantidade de valores "NaN" na coluna Rating após substituição: 0'

### Tratamento de valores "NaN"

Testamos se havia valores `NaN`em todas as colunas do dataframe e vimos que as seguintes colunas possuiam as seguintes quantidades desses valores ruidosos:
- Year_of_Release: 269
- Critic_Score: 8578
- Critic_Count: 8578
- User_Score: 9125
- User_Count: 9125
- Name: 2
- Genre: 2
- Publisher: 54
- Developer: 6620
- Rating: 6766

Como as colunas `Year_of_Release`, `Name`, `Genre` e `Publisher` possuem poucos valores ruidosos comparado ao total de dados do dataframe (16.715 linhas), optamos por dropar esses registros da base.

Já as colunas `Critic_Score`, `Critic_Count`, `User_Score` e `User_Count` possuiam mais de 8000 registros ruidosos. Sendo assim optamos por substituir os valores `NaN` por `0`, pois consideramos que `Critic_Count` e `User_Count` com valores `NaN` significa que nenhum usuário avaliou aquele jogo e por isso o `Critic_Score` e `User_Score` possuiriam nota 0.

Além disso, como `Rating` é uma coluna de valores categóricos e possuia mais de 6000 registros ruidosos. Neste caso, consideramos que um registro de rating `NaN` seria equivalente ao rating `E`.

Por fim, na coluna `Developer` consideramos que os registros que possuiam `NaN` como valor seria substituido pelo valor da `Publisher`.

In [124]:
df['Year_of_Release'] = df['Year_of_Release'].apply(lambda x : int(x))
df.drop(labels=['Critic_Count', 'User_Count'], axis=1, inplace=True)
df

Name Platform  \
0                                             Wii Sports      Wii   
1                                      Super Mario Bros.      NES   
2                                         Mario Kart Wii      Wii   
3                                      Wii Sports Resort      Wii   
4                               Pokemon Red/Pokemon Blue       GB   
5                                                 Tetris       GB   
6                                  New Super Mario Bros.       DS   
7                                               Wii Play      Wii   
8                              New Super Mario Bros. Wii      Wii   
9                                              Duck Hunt      NES   
10                                            Nintendogs       DS   
11                                         Mario Kart DS       DS   
12                           Pokemon Gold/Pokemon Silver       GB   
13                                               Wii Fit      Wii   
14                                    Kinect Adventures!     X360   
15                                          Wii Fit Plus      Wii   
16                                    Grand Theft Auto V      PS3   
17                         Grand Theft Auto: San Andreas      PS2   
18                                     Super Mario World     SNES   
19          Brain Age: Train Your Brain in Minutes a Day       DS   
20                         Pokemon Diamond/Pokemon Pearl       DS   
21                                      Super Mario Land       GB   
22                                   Super Mario Bros. 3      NES   
23                                    Grand Theft Auto V     X360   
24                           Grand Theft Auto: Vice City      PS2   
25                         Pokemon Ruby/Pokemon Sapphire      GBA   
26           Brain Age 2: More Training in Minutes a Day       DS   
27                           Pokemon Black/Pokemon White       DS   
28                                Gran Turismo 3: A-Spec      PS2   
29                        Call of Duty: Modern Warfare 3     X360   
...                                                  ...      ...   
16689              Help Wanted: 50 Wacky Jobs (jp sales)      Wii   
16690  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii   
16691                              Neo Angelique Special      PSP   
16692                      Outdoors Unleashed: Africa 3D      3DS   
16693                                          Real Rode      PS2   
16694                                     Pony Friends 2       PC   
16695                       Dynasty Warriors: Eiketsuden      PS3   
16696                  Metal Gear Solid V: Ground Zeroes       PC   
16697                                  PGA European Tour      N64   
16698                                     Carmageddon 64      N64   
16699                                    Planet Monsters      GBA   
16700                                             Breach       PC   
16701                                   Bust-A-Move 3000       GC   
16702                                   Mega Brain Boost       DS   
16703                              The Longest 5 Minutes      PSV   
16704                           Mezase!! Tsuri Master DS       DS   
16705  Eiyuu Densetsu: Sora no Kiseki Material Collec...      PSP   
16706                            STORM: Frontline Nation       PC   
16707                                   Strawberry Nauts      PSV   
16708                                           Plushees       DS   
16709                                            15 Days       PC   
16710                      Men in Black II: Alien Escape       GC   
16711                                   Aiyoku no Eustia      PSV   
16712                 Woody Woodpecker in Crazy Castle 5      GBA   
16713   SCORE International Baja 1000: The Official Game      PS2   
16714                      Samurai Warriors: Sanada Maru      PS3   
16715                                   LMA Manage

Convertemos os valores da coluna `Year_of_Release` de float para int.
Consideramos que as colunas `Critic_Count` e `User_Count` são irrelevantes dentre as hipóteses levantadas, e por isso decidimos dropálas da base.

## Normalização

Normalizamos os valores de vendas de acordo com o maior valor de cada coluna em específica `value/maxValue`.

In [125]:
max_NA_Sales = df['NA_Sales'].max()
print(max_NA_Sales)
max_EU_Sales = df['EU_Sales'].max()
print(max_EU_Sales)
max_JP_Sales = df['JP_Sales'].max()
print(max_JP_Sales)
max_Other_Sales = df['Other_Sales'].max()
print(max_Other_Sales)
max_Global_Sales = df['Global_Sales'].max()
print(max_Global_Sales)

41.36
28.96
10.22
10.57
82.53


In [126]:
for i in df.columns[5:]:
    if(i == 'NA_Sales'):
        df[i] = df[i].apply(lambda x : x/max_NA_Sales)
    elif(i == 'JP_Sales'):
        df[i] = df[i].apply(lambda x : x/max_JP_Sales)
    elif(i == 'EU_Sales'):
        df[i] = df[i].apply(lambda x : x/max_EU_Sales)
    elif(i == 'Other_Sales'):
        df[i] = df[i].apply(lambda x : x/max_Other_Sales)
    elif(i == 'Global_Sales'):
        df[i] = df[i].apply(lambda x : x/max_Global_Sales)
df

Name Platform  \
0                                             Wii Sports      Wii   
1                                      Super Mario Bros.      NES   
2                                         Mario Kart Wii      Wii   
3                                      Wii Sports Resort      Wii   
4                               Pokemon Red/Pokemon Blue       GB   
5                                                 Tetris       GB   
6                                  New Super Mario Bros.       DS   
7                                               Wii Play      Wii   
8                              New Super Mario Bros. Wii      Wii   
9                                              Duck Hunt      NES   
10                                            Nintendogs       DS   
11                                         Mario Kart DS       DS   
12                           Pokemon Gold/Pokemon Silver       GB   
13                                               Wii Fit      Wii   
14                                    Kinect Adventures!     X360   
15                                          Wii Fit Plus      Wii   
16                                    Grand Theft Auto V      PS3   
17                         Grand Theft Auto: San Andreas      PS2   
18                                     Super Mario World     SNES   
19          Brain Age: Train Your Brain in Minutes a Day       DS   
20                         Pokemon Diamond/Pokemon Pearl       DS   
21                                      Super Mario Land       GB   
22                                   Super Mario Bros. 3      NES   
23                                    Grand Theft Auto V     X360   
24                           Grand Theft Auto: Vice City      PS2   
25                         Pokemon Ruby/Pokemon Sapphire      GBA   
26           Brain Age 2: More Training in Minutes a Day       DS   
27                           Pokemon Black/Pokemon White       DS   
28                                Gran Turismo 3: A-Spec      PS2   
29                        Call of Duty: Modern Warfare 3     X360   
...                                                  ...      ...   
16689              Help Wanted: 50 Wacky Jobs (jp sales)      Wii   
16690  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii   
16691                              Neo Angelique Special      PSP   
16692                      Outdoors Unleashed: Africa 3D      3DS   
16693                                          Real Rode      PS2   
16694                                     Pony Friends 2       PC   
16695                       Dynasty Warriors: Eiketsuden      PS3   
16696                  Metal Gear Solid V: Ground Zeroes       PC   
16697                                  PGA European Tour      N64   
16698                                     Carmageddon 64      N64   
16699                                    Planet Monsters      GBA   
16700                                             Breach       PC   
16701                                   Bust-A-Move 3000       GC   
16702                                   Mega Brain Boost       DS   
16703                              The Longest 5 Minutes      PSV   
16704                           Mezase!! Tsuri Master DS       DS   
16705  Eiyuu Densetsu: Sora no Kiseki Material Collec...      PSP   
16706                            STORM: Frontline Nation       PC   
16707                                   Strawberry Nauts      PSV   
16708                                           Plushees       DS   
16709                                            15 Days       PC   
16710                      Men in Black II: Alien Escape       GC   
16711                                   Aiyoku no Eustia      PSV   
16712                 Woody Woodpecker in Crazy Castle 5      GBA   
16713   SCORE International Baja 1000: The Official Game      PS2   
16714                      Samurai Warriors: Sanada Maru      PS3   
16715                                   LMA Manage

In [127]:
print("qtd possíveis valores para plataforma: {}".format(len(df['Platform'].value_counts())))
print("qtd possíveis valores para genero: {}".format(len(df['Genre'].value_counts())))
print("qtd possíveis valores para publisher: {}".format(len(df['Publisher'].value_counts())))
print("qtd possíveis valores para developer: {}".format(len(df['Developer'].value_counts())))
print("qtd possíveis valores para rating: {}".format(len(df['Rating'].value_counts())))

qtd possíveis valores para plataforma: 31
qtd possíveis valores para genero: 12
qtd possíveis valores para publisher: 579
qtd possíveis valores para developer: 1979
qtd possíveis valores para rating: 8


In [128]:
gen_le = LabelEncoder()
gen_labels = gen_le.fit_transform(df['Genre'])
df['Genre_Labels'] = gen_labels

plat_le = LabelEncoder()
platform_labels = plat_le.fit_transform(df['Platform'])
df['Platform_Labels'] = platform_labels

rat_le = LabelEncoder()
rating_labels = rat_le.fit_transform(df['Rating'])
df['Rating_Labels'] = rating_labels

df_sub = df[['Name', 'Genre', 'Genre_Labels',  
                       'Platform', 'Platform_Labels','Rating','Rating_Labels']]

df_sub

Name         Genre  \
0                                             Wii Sports        Sports   
1                                      Super Mario Bros.      Platform   
2                                         Mario Kart Wii        Racing   
3                                      Wii Sports Resort        Sports   
4                               Pokemon Red/Pokemon Blue  Role-Playing   
5                                                 Tetris        Puzzle   
6                                  New Super Mario Bros.      Platform   
7                                               Wii Play          Misc   
8                              New Super Mario Bros. Wii      Platform   
9                                              Duck Hunt       Shooter   
10                                            Nintendogs    Simulation   
11                                         Mario Kart DS        Racing   
12                           Pokemon Gold/Pokemon Silver  Role-Playing   
13                                               Wii Fit        Sports   
14                                    Kinect Adventures!          Misc   
15                                          Wii Fit Plus        Sports   
16                                    Grand Theft Auto V        Action   
17                         Grand Theft Auto: San Andreas        Action   
18                                     Super Mario World      Platform   
19          Brain Age: Train Your Brain in Minutes a Day          Misc   
20                         Pokemon Diamond/Pokemon Pearl  Role-Playing   
21                                      Super Mario Land      Platform   
22                                   Super Mario Bros. 3      Platform   
23                                    Grand Theft Auto V        Action   
24                           Grand Theft Auto: Vice City        Action   
25                         Pokemon Ruby/Pokemon Sapphire  Role-Playing   
26           Brain Age 2: More Training in Minutes a Day        Puzzle   
27                           Pokemon Black/Pokemon White  Role-Playing   
28                                Gran Turismo 3: A-Spec        Racing   
29                        Call of Duty: Modern Warfare 3       Shooter   
...                                                  ...           ...   
16689              Help Wanted: 50 Wacky Jobs (jp sales)    Simulation   
16690  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...        Racing   
16691                              Neo Angelique Special     Adventure   
16692                      Outdoors Unleashed: Africa 3D        Sports   
16693                                          Real Rode     Adventure   
16694                                     Pony Friends 2    Simulation   
16695                       Dynasty Warriors: Eiketsuden        Action   
16696                  Metal Gear Solid V: Ground Zeroes        Action   
16697                                  PGA European Tour        Sports   
16698                                     Carmageddon 64        Action   
16699                                    Planet Monsters        Action   
16700                                             Breach       Shooter   
16701                                   Bust-A-Move 3000        Puzzle   
16702                                   Mega Brain Boost        Puzzle   
16703                              The Longest 5 Minutes        Action   
16704                           Mezase!! Tsuri Master DS        Sports   
16705  Eiyuu Densetsu: Sora no Kiseki Material Collec...  Role-Playing   
16706                            STORM: Frontline Nation      Strategy   
16707                                   Strawberry Nauts     Adventure   
16708                                           Plushees    Simulation   
16709                                            15 Days     Adventure   
16710                      Men in Black II: Alien Escape       Shooter   
16711                                   Aiyoku no Eustia

In [129]:
gen_ohe = OneHotEncoder()
gen_feature_arr = gen_ohe.fit_transform(df[['Genre_Labels']]).toarray()
gen_feature_labels = list(gen_le.classes_)
gen_features = pd.DataFrame(gen_feature_arr, columns=gen_feature_labels)

plat_ohe = OneHotEncoder()
plat_feature_arr = plat_ohe.fit_transform(df[['Platform_Labels']]).toarray()
plat_feature_labels = list(plat_le.classes_)
plat_features = pd.DataFrame(plat_feature_arr, columns=plat_feature_labels)

rat_ohe = OneHotEncoder()
rat_feature_arr = rat_ohe.fit_transform(df[['Rating_Labels']]).toarray()
rat_feature_labels = list(rat_le.classes_)
rat_features = pd.DataFrame(rat_feature_arr, columns=rat_feature_labels)

df = pd.concat([df.reset_index(drop=True), gen_features.reset_index(drop=True),plat_features.reset_index(drop=True), rat_features.reset_index(drop=True)], axis=1)
df

Name Platform  \
0                                             Wii Sports      Wii   
1                                      Super Mario Bros.      NES   
2                                         Mario Kart Wii      Wii   
3                                      Wii Sports Resort      Wii   
4                               Pokemon Red/Pokemon Blue       GB   
5                                                 Tetris       GB   
6                                  New Super Mario Bros.       DS   
7                                               Wii Play      Wii   
8                              New Super Mario Bros. Wii      Wii   
9                                              Duck Hunt      NES   
10                                            Nintendogs       DS   
11                                         Mario Kart DS       DS   
12                           Pokemon Gold/Pokemon Silver       GB   
13                                               Wii Fit      Wii   
14                                    Kinect Adventures!     X360   
15                                          Wii Fit Plus      Wii   
16                                    Grand Theft Auto V      PS3   
17                         Grand Theft Auto: San Andreas      PS2   
18                                     Super Mario World     SNES   
19          Brain Age: Train Your Brain in Minutes a Day       DS   
20                         Pokemon Diamond/Pokemon Pearl       DS   
21                                      Super Mario Land       GB   
22                                   Super Mario Bros. 3      NES   
23                                    Grand Theft Auto V     X360   
24                           Grand Theft Auto: Vice City      PS2   
25                         Pokemon Ruby/Pokemon Sapphire      GBA   
26           Brain Age 2: More Training in Minutes a Day       DS   
27                           Pokemon Black/Pokemon White       DS   
28                                Gran Turismo 3: A-Spec      PS2   
29                        Call of Duty: Modern Warfare 3     X360   
...                                                  ...      ...   
16382              Help Wanted: 50 Wacky Jobs (jp sales)      Wii   
16383  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii   
16384                              Neo Angelique Special      PSP   
16385                      Outdoors Unleashed: Africa 3D      3DS   
16386                                          Real Rode      PS2   
16387                                     Pony Friends 2       PC   
16388                       Dynasty Warriors: Eiketsuden      PS3   
16389                  Metal Gear Solid V: Ground Zeroes       PC   
16390                                  PGA European Tour      N64   
16391                                     Carmageddon 64      N64   
16392                                    Planet Monsters      GBA   
16393                                             Breach       PC   
16394                                   Bust-A-Move 3000       GC   
16395                                   Mega Brain Boost       DS   
16396                              The Longest 5 Minutes      PSV   
16397                           Mezase!! Tsuri Master DS       DS   
16398  Eiyuu Densetsu: Sora no Kiseki Material Collec...      PSP   
16399                            STORM: Frontline Nation       PC   
16400                                   Strawberry Nauts      PSV   
16401                                           Plushees       DS   
16402                                            15 Days       PC   
16403                      Men in Black II: Alien Escape       GC   
16404                                   Aiyoku no Eustia      PSV   
16405                 Woody Woodpecker in Crazy Castle 5      GBA   
16406   SCORE International Baja 1000: The Official Game      PS2   
16407                      Samurai Warriors: Sanada Maru      PS3   
16408                                   LMA Manage

In [ ]:
df.to_csv(r'Path where you want to store the exported CSV file\File Name.csv')